File: LeetCode_Algorithm/235/235.js
function TreeNode(val) {
  this.val = val
  this.left = this.right = null

}
function makeTree(root, idx) {
  var rootNode = new TreeNode(root[idx])
  if (2*idx+1 < root.length) {
    rootNode.left = makeTree(root, 2*idx+1)
  }
  if (2*idx+2 < root.length) {
    rootNode.right = makeTree(root, 2*idx+2)
  }
  return rootNode
}

var lowestCommonAncestor = function(root, p, q) {
  rootNode = makeTree(root, 0)
  // O(n) 101ms 76%
  while (rootNode != null) {
    if (p < rootNode.val && q < rootNode.val) {
      rootNode = rootNode.left
    }
    else if (rootNode.val < p && rootNode.val < q) {
      rootNode = rootNode.right
    }
    else {
      return rootNode.val
    }
  }

};

console.log(lowestCommonAncestor([6,2,8,0,4,7,9,null,null,3,5], 2, 4))

File: LeetCode_Algorithm/235/235.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        rootNode = self.makeTree(root, 0)
        # O(n) O(1) 145ms 17% iterative
        # while rootNode:
        #     rootnode_val = rootNode.val
        #     if p > rootnode_val and q > rootnode_val:
        #         rootNode = rootNode.right
        #     elif p < rootnode_val and q < rootnode_val:
        #         rootNode = rootNode.left
        #     else:
        #         return rootNode

        # sol2 O(n) O(n)(재귀 스택) recursive 89ms 76%
        return self.solve(rootNode, p, q)
    def solve(self, root, p, q):
        if p > root.val and q > root.val:
            return self.solve(root.right, p, q)
        elif p < root.val and q < root.val:
            return self.solve(root.left, p, q)
        else:
            return root.val

    def makeTree(self, root, idx):
        rootNode = TreeNode(root[idx])
        if 2*idx+1 < len(root): rootNode.left = self.makeTree(root, 2*idx+1)
        if 2*idx+2 < len(root): rootNode.right = self.makeTree(root, 2*idx+2)
        return rootNode



print(Solution().lowestCommonAncestor([6,2,8,0,4,7,9,None,None,3,5], 2,8))